File: dataload/owl2json/src/main/java/uk/ac/ebi/owl2json/xrefs/Bioregistry.java

package uk.ac.ebi.owl2json.xrefs;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.client.config.RequestConfig;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClientBuilder;

import javax.annotation.RegEx;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.regex.Pattern;


/**
 * The Bioregistry (https://bioregistry.io) is an open source, community curated registry
 * of prefixes for biomedical ontologies/vocabularies and their associated metadata. It
 * can be used to look generate links for prefixes found in xrefs and other components of
 * ontologies.
 *
 * Source code and data is availale under CC0/MIT licenses at https://github.com/biopragmatics/bioregistry
 */
public class Bioregistry {

    Gson gson = new Gson();

    public String registryUrl;

    JsonObject theRegistry;
    Map<String, JsonObject> prefixToDatabase = new HashMap<>();
    Map<String, Pattern> patterns = new HashMap<>();

    public Bioregistry() {
        this("https://raw.githubusercontent.com/biopragmatics/bioregistry/main/exports/registry/registry.json");
    }

    public Bioregistry(String jsonUrl){

        this.registryUrl = jsonUrl;

        try {
            theRegistry = urlToJson(jsonUrl).getAsJsonObject();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        for(var entry : theRegistry.entrySet()) {
            
            JsonObject db = entry.getValue().getAsJsonObject();

            // The key is the canonical Bioregistry prefix, always lowercase
            prefixToDatabase.put(norm(entry.getKey()), db);

            // The preferred prefix can have various capitalization,
            // usually is same as canonical
            prefixToDatabase.put(norm(db.get("preferred_prefix").getAsString()), db);

            JsonElement synonyms = db.get("synonyms");

            if(synonyms != null) {
                for(JsonElement synonym : synonyms.getAsJsonArray()) {
                    prefixToDatabase.put(norm(synonym.getAsString()), db);
                }
            }
        }

    }

    public String getRegistryUrl() {
        return  registryUrl;
    }

    public String getUrlForId(String databaseId, String id) {

        JsonObject db = prefixToDatabase.get(norm(databaseId));

        if(db == null)
            return null;

        if(id == null)
            return null;

        JsonElement patternObj = db.get("pattern");

        if(patternObj == null)
            return null;

        Pattern pattern = getPattern(patternObj.getAsString());

        if(!pattern.matcher(id).matches()) {
            return null;
        }

        JsonElement uriFormat = db.get("uri_format");

        if(uriFormat == null) {
            return null;
        }
            
        return uriFormat.getAsString().replace("$1", id);
    }
    
    private static String norm(String s) {
        // see https://github.com/biopragmatics/bioregistry/blob/a7424ef4a0d22eaca61d3a86c6175e2059e9c855/src/bioregistry/utils.py#L128-L133
        s = s.toLowerCase(Locale.ROOT);
        s = s.replace(".", "");
        s = s.replace("-", "");
        s = s.replace("_", "");
        s = s.replace("/", "");
        return s;
    }

    private Pattern getPattern(String patternStr) {

        Pattern found = patterns.get(patternStr);

        if(found != null) {
            return found;
        }

        Pattern pattern = Pattern.compile(patternStr);
        patterns.put(patternStr, pattern);
        return pattern;
    }

    private JsonElement urlToJson(String url) throws IOException {

        RequestConfig config = RequestConfig.custom()
                .setConnectTimeout(5000)
                .setConnectionRequestTimeout(5000)
                .setSocketTimeout(5000).build();

        CloseableHttpClient client = HttpClientBuilder.create().setDefaultRequestConfig(config).build();

        HttpGet request = new HttpGet(url);
        HttpResponse response = client.execute(request);
        HttpEntity entity = response.getEntity();
        if (entity != null) {
            return new JsonParser().parse(new InputStreamReader(entity.getContent()));
        } else {
            throw new RuntimeException("bioregistry response was null");
        }
    }

}

File: dataload/linker/link/src/bioregistry.rs
use regex::Regex;
use serde::Deserialize;
use std::collections::HashMap;
use std::sync::LazyLock;

/// The Bioregistry (https://bioregistry.io) is an open source, community curated registry
/// of prefixes for biomedical ontologies/vocabularies and their associated metadata.
/// It can be used to look generate links for prefixes found in xrefs and other components of
/// ontologies.
///
/// Source code and data is available under CC0/MIT licenses at https://github.com/biopragmatics/bioregistry

const DEFAULT_REGISTRY_URL: &str = "https://raw.githubusercontent.com/biopragmatics/bioregistry/main/exports/registry/registry.json";

#[derive(Debug, Deserialize)]
struct RegistryEntry {
    #[serde(default)]
    preferred_prefix: Option<String>,
    #[serde(default)]
    synonyms: Option<Vec<String>>,
    #[serde(default)]
    uri_format: Option<String>,
    #[serde(default)]
    pattern: Option<String>,
}

pub struct Bioregistry {
    registry_url: String,
    prefix_to_database: HashMap<String, DatabaseEntry>,
    /// Maps URI prefix -> canonical database ID (sorted by length descending for longest match first)
    iri_prefix_to_database: Vec<(String, String)>,
    patterns: HashMap<String, Regex>,
}

#[derive(Clone)]
struct DatabaseEntry {
    uri_format: Option<String>,
    pattern: Option<String>,
}

impl Bioregistry {
    pub fn new() -> Result<Self, Box<dyn std::error::Error>> {
        Self::with_url(DEFAULT_REGISTRY_URL)
    }

    pub fn with_url(json_url: &str) -> Result<Self, Box<dyn std::error::Error>> {
        let response = ureq::get(json_url).call()?;
        let registry: HashMap<String, RegistryEntry> = response.into_json()?;
        
        let mut prefix_to_database: HashMap<String, DatabaseEntry> = HashMap::new();
        // Use a HashMap to deduplicate - later entries with same prefix will overwrite earlier ones
        let mut iri_prefix_map: HashMap<String, String> = HashMap::new();
        
        // Sort registry keys to get consistent order (like Java's iteration over JSON object)
        let mut registry_keys: Vec<_> = registry.keys().collect();
        registry_keys.sort();
        
        for key in registry_keys {
            let entry = &registry[key];
            let db_entry = DatabaseEntry {
                uri_format: entry.uri_format.clone(),
                pattern: entry.pattern.clone(),
            };
            
            // The key is the canonical Bioregistry prefix, always lowercase
            prefix_to_database.insert(norm(key), db_entry);
            
            // The preferred prefix can have various capitalization
            if let Some(ref preferred) = entry.preferred_prefix {
                let db_entry = DatabaseEntry {
                    uri_format: entry.uri_format.clone(),
                    pattern: entry.pattern.clone(),
                };
                prefix_to_database.insert(norm(preferred), db_entry);
            }
            
            // Handle synonyms
            if let Some(ref synonyms) = entry.synonyms {
                for synonym in synonyms {
                    let db_entry = DatabaseEntry {
                        uri_format: entry.uri_format.clone(),
                        pattern: entry.pattern.clone(),
                    };
                    prefix_to_database.insert(norm(synonym), db_entry);
                }
            }
            
            // Build IRI prefix mapping - later entries with same uri_prefix will overwrite
            if let Some(ref uri_format) = entry.uri_format {
                if uri_format.ends_with("$1") {
                    let uri_prefix = &uri_format[..uri_format.len() - 2];
                    iri_prefix_map.insert(uri_prefix.to_string(), key.to_string());
                }
            }
        }
        
        // Convert to Vec and sort by length descending for longest match first
        let mut iri_prefix_to_database: Vec<(String, String)> = iri_prefix_map.into_iter().collect();
        iri_prefix_to_database.sort_by(|a, b| {
            b.0.len().cmp(&a.0.len()).then_with(|| a.0.cmp(&b.0))
        });
        
        Ok(Self {
            registry_url: json_url.to_string(),
            prefix_to_database,
            iri_prefix_to_database,
            patterns: HashMap::new(),
        })
    }

    pub fn get_registry_url(&self) -> &str {
        &self.registry_url
    }

    pub fn get_url_for_id(&mut self, database_id: &str, id: &str) -> Option<String> {
        // Clone the database entry to avoid borrow conflicts
        let db = self.prefix_to_database.get(&norm(database_id))?.clone();
        
        // Java returns null if there's no pattern - we need to match this behavior
        let pattern_str = db.pattern.as_ref()?;
        
        let pattern = self.get_pattern(pattern_str);
        // We need to check if the pattern matches the ENTIRE string (like Java's matches())
        // Using find() and checking that it covers the whole string
        if let Some(m) = pattern.find(id) {
            if m.start() != 0 || m.end() != id.len() {
                return None;
            }
        } else {
            return None;
        }
        
        let uri_format = db.uri_format.as_ref()?;
        Some(uri_format.replace("$1", id))
    }

    pub fn get_curie_for_url(&self, url: &str) -> Option<String> {
        for (prefix, database_id) in &self.iri_prefix_to_database {
            if url.starts_with(prefix) {
                let local_unique_identifier = &url[prefix.len()..];
                return Some(format!("{}:{}", database_id, local_unique_identifier));
            }
        }
        None
    }

    fn get_pattern(&mut self, pattern_str: &str) -> &Regex {
        if !self.patterns.contains_key(pattern_str) {
            let pattern = Regex::new(pattern_str).unwrap_or_else(|_| Regex::new(".*").unwrap());
            self.patterns.insert(pattern_str.to_string(), pattern);
        }
        self.patterns.get(pattern_str).unwrap()
    }
}

/// Normalize a prefix string according to bioregistry conventions
fn norm(s: &str) -> String {
    s.to_lowercase()
        .replace('.', "")
        .replace('-', "")
        .replace('_', "")
        .replace('/', "")
}

// Check if a string looks like a CURIE
// Note: [A-z] matches A-Z, [, \, ], ^, _, `, a-z (ASCII 65-122)
// This matches Java's behavior exactly
static CURIE_PATTERN: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"^[A-z0-9]+:[A-z0-9-]+$").unwrap()
});

pub fn is_curie(s: &str) -> bool {
    CURIE_PATTERN.is_match(s)
}
